### PR TITLE
Part two of how to loadtest with Oauth and Firefox Account.

### PR DESCRIPTION
**Pull request by Natim**
The draft is available here: http://www.servicedenuages.fr/drafts/how-to-load-test-a-oauth-protected-http-api-part-2-en.html
feedback @almet @rfk @tarekziade ?

### CODE REVIEW

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 1]
```diff
@@ -0,0 +1,264 @@
+How to load test a OAuth-protected HTTP API? (Part 2)
+#####################################################
+
+:lang: en
+:date: 2016-01-05
+:url: load-testing-a-http-api-which-uses-oauth-for-authentication-part-2
+:summary: This is part 2 of a series on loadtesting oauth APIs explaining how to test the performance of an HTTP API when its access depends on an external OAuth service.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: You can summarize a bit here. "How to loadtest OAuth APIs, part 2: how to generate FxA tokens."

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+This is part 2 of a series on loadtesting oauth APIs.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: OAuth

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+Introduction
+============
+
+`In the previous article </en/load-testing-a-http-api-which-uses-oauth-for-authentication>`_
+we have seen that:
+
+- We would create the OAuth Bearer Token before the load-test start.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: "We can create the OAuth".

Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: start_s_

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+- We can use environment variables to use the token in the load-test.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: to specify the token in the loadtest.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+In this article we will have a closer look at how we can generate
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: At how to generate

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+tokens for Firefox Account and pass them to the load-test.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Firefox Account**s**

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+
+PyFxA: the Python Firefox Account client
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: The Firefox Accounts python client.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+========================================
+
+`PyFxA <https://github.com/mozilla/PyFxA>`_ is a Python library that
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: a python library providing helpers*

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+aims providing helpers to interact with the Firefox Account ecosystem.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Account**s**

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+
+Using PyFxA with httpie
+-----------------------
+
+A while ago we added support for 
+`Requests Authentication <http://docs.python-requests.org/en/latest/user/authentication/>`_
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Consider reversing the phrasing: "To ease the use of FxA bearer tokens, we've created a authentication policy for the requests library, which can also be used with the httpie client".

I would specify directly a link to this authentication policy.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+to ease the use of Firefox Account Bearer Token with
+`requests <http://docs.python-requests.org/en/latest/>`_ and 
+`httpie <http://httpie.org>`_.
+
+With this plugin, we can basically provide our user credentials and
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: s/basically//

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+the plugin will build a valid Bearer Token for our request:
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: and a valid bearer token will be built for our request.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+.. code-block:: bash
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst:45]: Here is what it looks like with HTTPie.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+
+    http https://profile.accounts.firefox.com/v1/profile \
+        --auth-type fxa-bearer \
+        --auth "email@domain.tld:password"
+
+
+Using fxa-client to generate Bearer Tokens
+------------------------------------------
+
+The code we wrote at the time was quite similar to what we were trying
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst:55]: "At the time"? which time are you talking about?

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+to do for the load-test setup: Generating a Bearer Token for a
+given user.
+
+We ended-up building a CLI tool called ``fxa-client`` that is able to
+generate a BASH file that export some environment variables:
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: export_s_

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+    # ---- BEARER TOKEN INFO ----
+    # User: my-app-6318a65dde1efc2f4c3f7b4e6cb33188@restmail.net
+    # Scopes: profile
+    # Account: https://api-accounts.stage.mozaws.net/v1
+    # Oauth: https://oauth.stage.mozaws.net/v1
+    # Client ID: 5882386c6d801776
+    # ---------------------------
+    export OAUTH_BEARER_TOKEN="90abc87ed1621ee504c1252ed382abc8269d1abc29f2ff87cc5e25f00249fabc"
+
+If we want to reuse the same account multiple time and create it only
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: multiple time**s**

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+if needed, we can provide the user-salt (as a valid base64 string):
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: You need to rephrase this sentence: "To avoid creating extraneous user accounts, it is possible to specify a user-salt [...]"

You need to explain what the user salt would be used for.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+
+.. code-block:: bash
+
+    fxa-client --create --bearer --prefix my-app --user-salt MySalt==
+
+
+Using fxa-client to work with BrowserID assertion
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: assertions.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+-------------------------------------------------
+
+Even if we tend to switch from BrowserID assertion to Bearer Tokens
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: assertions

Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: You need to be consistent with how you spell "Bearer Tokens". either put it in Camel Case everywhere, or in lower case.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+for our new services, if we need to work with Firefox Sync or Firefox
+Hello we may need to provide a BrowserID assertion to log into these
+services.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: This sentence is hard to follow for people not aware of what kind of auth Sync and Hello are using.

File: content/2016.01.loadtesting.with-oauth-part-2.rst (hunk 19 above)
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Maybe rephrase: "We are currently relying on OAuth2 Bearer Tokens for our new services. However, some of the old services (Sync, Firefox Hello) still rely on BrowserID assertions. As such, we need to provide a valid BrowserID assertion to log into these services".

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+In that case ``fxa-client`` provides few specific attributes:
+
+.. code-block:: bash
+
+    optional arguments:
+      --browserid, --bid    Generate a BrowserID assertion
+      --audience AUDIENCE   Firefox BrowserID assertion audience.
+      --duration DURATION   Firefox BrowserID assertion duration.
+
+It works just like before with Bearer Tokens but generate a BrowserID
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: The script work exactly in the same way than the one for OAuth2 Bearer Tokens, except it generates a BrowserID assertion

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+
+    # ---- BROWSER ID ASSERTION INFO ----
+    # User: my-app-b82d4afaf57cb856ccc04a58a07ce80f@restmail.net
+    # Audience: https://loop.stage.mozaws.net
+    # Account: https://api-accounts.stage.mozaws.net/v1
+    # ------------------------------------
+    export FXA_BROWSERID_ASSERTION="eyJhbGciOiJSUzI1NiJ9...hIQ9vrkqA"
+    export FXA_CLIENT_STATE="828aef3bc68ac0bde10f3d4b93303088"
+
+And then the assertion could be used like that:
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: s/could/can

Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Can be used is the `Authorization` header of the request::

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+        Authorization:"BrowserID eyJhbGciOiJSUzI1NiJ9...hIQ9vrkqA"
+
+
+Using fxa-client to configure a loadtest
+========================================
+
+Now that we have a quite simple way to generate Bearer Tokens, how can
+we plug that with our load-tests?
+
+I could not find a better way than creating a bash file that export
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: export**s**

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+the environment variable and then loading it in the shell that will
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: variable**s**

Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: The end of the sentence " and then loading it in the shell that will run the loadtests" seems superfluous.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
+
+Something like:
+
+.. code-block:: bash
+
+    fxa-client --create --bearer --user-salt MySalt== --out loadtest-fxa-config.sh
+    source loadtest-fxa-config.sh
+    docker run -e OAUTH_BEARER_TOKEN="${OAUTH_BEARER_TOKEN}" loadtest
+
+After we have sourced the ``loadtest-fxa-config.sh`` file the env
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: After sourcing the

Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: After sourcing the [...]**,** the env

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+variables are exposed so that we can read them from any program
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: are exposed. Any program (regardless of the language) could read them if needed.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+
+
+How to install fxa-client?
+==========================
+
+Right now ``fxa-client`` is still a work in progress and didn't land yet
+in the last release of PyFxA. It doesn't mean we cannot use it
+already.
+
+To do so, just install the ``loadtest-tools`` branch of PyFxA:
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: PyFxA repository.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+
+.. code-block:: bash
+
+    pip install https://github.com/mozilla/PyFxA/archive/loadtest-tools.zip
+
+
+What's next?
+============
+
+Multiple account loadtest
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: accounts*

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+That's about it. I hope that after reading this article, you are not
+afraid anymore of load testing Firefox Account OAuth-based services!
+
+Take aways:
+
+- ``fxa-client`` let us generate a bash script with our user credentials.
+- This bash script can be loaded before running our load-test to expose
+  user credentials to a load-test script.
+
+Do not hesitate to reach to us if you have any questions or suggestions.
```
Comment (almet) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: to reach us.

File: content/2016.01.loadtesting.with-oauth-part-2.rst (hunk 14 above)
Comment (Natim) [inline content/2016.01.loadtesting.with-oauth-part-2.rst:55]: This is a reference to the `A while ago` sentence. (at the time : à l'époque)

File: content/2016.01.loadtesting.with-oauth-part-2.rst (hunk 3 above)
Comment (Natim) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: - I am using would because not only we can but **we want to do so**
- I wanted to use the load-test start (the noun start and not the verb starts)

Comment (rfk) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: I suggest "should" may be better than "would" in this case, to capture the notion that it's the right thing to do based on discussion the previous article.

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 30]
```diff
@@ -0,0 +1,269 @@
...
+
+
+Using PyFxA with httpie
+-----------------------
+
+To ease the use of Firefox Accounts Bearer Tokens, we have created, a
+while ago, an
+`authentication policy <https://github.com/mozilla/PyFxA/blob/master/fxa/plugins/requests.py#L115-L158>`_
+for the `requests library <http://docs.python-requests.org/en/latest/>`_, which
+can all be used with the `httpie <http://httpie.org>`_ client.
```
Comment (rfk) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: "can also be used"?

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 31]
```diff
@@ -0,0 +1,269 @@
...
+    http https://profile.accounts.firefox.com/v1/profile \
+        --auth-type fxa-bearer \
+        --auth "email@domain.tld:password"
+
+
+Using fxa-client to generate Bearer Tokens
+------------------------------------------
+
+The code we wrote at the time was quite similar to what we were trying
+to do for the load-test setup: Generating a Bearer Token for a
```
Comment (tarekziade) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: load test

Comment (tarekziade) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: s/Generating/generating

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 32]
```diff
@@ -0,0 +1,269 @@
...
+    # User: my-app-6318a65dde1efc2f4c3f7b4e6cb33188@restmail.net
+    # Scopes: profile
+    # Account: https://api-accounts.stage.mozaws.net/v1
+    # Oauth: https://oauth.stage.mozaws.net/v1
+    # Client ID: 5882386c6d801776
+    # ---------------------------
+    export OAUTH_BEARER_TOKEN="90abc87ed1621ee504c1252ed382abc8269d1abc29f2ff87cc5e25f00249fabc"
+
+To avoid creating extraneous user accounts, it is possible to specify
+a user-salt (as a base64 string) that will always generate the same
```
Comment (tarekziade) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: user salt

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 33]
```diff
@@ -0,0 +1,269 @@
...
+Management Firefox Accounts environments
+----------------------------------------
+
+If we want to generate tokens for other Firefox Accounts environment,
+we would need to provide ``--account-server`` and ``--oauth-server``
+which default to the stage environment.
+
+We can find `all the available environment here <https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction#Firefox_Accounts_deployments>`_.
+
+However to ease the switch from on to the other we added a ``--env``
```
Comment (tarekziade) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: s/on/one

File: content/2016.01.loadtesting.with-oauth-part-2.rst [hunk 34]
```diff
@@ -0,0 +1,269 @@
...
+    export OAUTH_BEARER_TOKEN="90abc87ed1621ee504c1252ed382abc8269d1abc29f2ff87cc5e25f00249fabc,abc9087ed1621ee504c1252ed382abc8269d1abc29f2ff87cc5e25f00249fabc"
+
+
+Conclusion
+==========
+
+That's about it. I hope that after reading this article, you are not
+afraid anymore of load testing Firefox Accounts OAuth-based services!
+
+Take aways:
```
Comment (tarekziade) [inline content/2016.01.loadtesting.with-oauth-part-2.rst]: Take-aways
